src/syntaxtree.c

```diff
@@ -31,50 +31,101 @@
 #include <ctype.h>
 #include <stdarg.h>
 
+enum { MAX_ARITY = 4 };
+
 typedef Obj (*CompileFuncT)(Obj node, Expr expr);
 typedef Obj (*CompileArgT)(Expr expr);
+typedef Expr (*CodeFuncT)(Obj node);
+typedef Expr (*CodeArgT)(Obj node);
+
 
 typedef struct {
     const Char * argname;
     CompileArgT  argcomp;
+    CodeArgT     argcode;
+    Int          isStat;
 } ArgT;
 
 typedef struct {
     UInt1        tnum;
     CompileFuncT compile;
+    CodeFuncT    code;
     const Char * name;
-    UInt1        arity;
-    ArgT         args[4]; /* The maximum compiler arity is currently 4 */
+    UInt         arity;
+    ArgT args[MAX_ARITY]; /* The maximum compiler arity is currently 4 */
 } CompilerT;
 
 // We put compilers for statements and expressions into the same static array,
 // assuming that the set of their respective ids are disjoint.
 static const CompilerT Compilers[];
 #define COMPILER_ARITY(...) (sizeof((ArgT[]){ __VA_ARGS__ }) / sizeof(ArgT))
-#define COMPILER_GENERIC(tnum, compiler, ...)                                \
+#define COMPILER_GENERIC(tnum, compiler, coder, ...)                         \
     [tnum] = {                                                               \
-        tnum, compiler, #tnum, COMPILER_ARITY(__VA_ARGS__), { __VA_ARGS__ }  \
+        tnum,           compiler, coder, #tnum, COMPILER_ARITY(__VA_ARGS__), \
+        { __VA_ARGS__ }                                                      \
     }
 
-#define COMPILER(tnum, compiler) COMPILER_GENERIC(tnum, compiler)
+#define COMPILER(tnum, compiler, coder)                                      \
+    COMPILER_GENERIC(tnum, compiler, coder)
 
 #define COMPILER_(tnum, ...)                                                 \
-    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler, __VA_ARGS__)
+    COMPILER_GENERIC(tnum, SyntaxTreeDefaultCompiler,                        \
+                     SyntaxTreeDefaultCoder, __VA_ARGS__)
 
-#define ARG(name, func)                                                      \
+#define ARG(name, func, coder, isStat)                                       \
     {                                                                        \
-        name, func                                                           \
+        name, func, coder, isStat                                            \
     }
-#define ARG_(name) ARG(name, SyntaxTreeCompiler)
 
-#define ARGS(name) ARG(name, 0)
+#define ARG_EXPR(name, func, coder) ARG(name, func, coder, 0)
+// The corresponding ARG_STAT is not used
+
+#define ARG_STAT_(name)                                                      \
+    ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultStatCoder, 1)
+#define ARG_EXPR_(name)                                                      \
+    ARG(name, SyntaxTreeCompiler, SyntaxTreeDefaultExprCoder, 0)
 
+#define ARGS_STAT(name) ARG(name, 0, 0, 1)
+#define ARGS_EXPR(name) ARG(name, 0, 0, 0)
```

Comment:
> // ARGS_EXPR can be used instead of ARG_EXPR as the last argument to the COMPILER_
> // macro, and indicates that a variable number of arguments follow; all are supposed to be
> // expressions; and all of them are to be stored into a plist which is stored in resp. read
> // from the record entry <name>.